### PR TITLE
Refactoring filtering

### PR DESCRIPTION
**Pull request by stoeffel**
Small refactor to make modifying a ListWidget.List simpler.
I thought that it would still be nicer to not make `Modifiable.ListWidget` to bloated by also adding `data Filter`.

### CODE REVIEW

File: nri-log-explorer/nri-log-explorer.cabal [hunk 1]
```diff
@@ -29,6 +29,7 @@ source-repository head
 executable log-explorer
   main-is: Main.hs
   other-modules:
+      Modifiable
```
Comment (stoeffel) [inline nri-log-explorer/nri-log-explorer.cabal:32]: Comment in commit is actually wrong. We can just use the `Functor` instance of `ListWidget` for search.

File: nri-log-explorer/src/Modifiable.hs [hunk 2]
```diff
@@ -0,0 +1,60 @@
+module Modifiable
+  ( ListWidget,
+    init,
+    cons,
+    toListWidget,
+    setListWidget,
+    filter,
+    reset,
+  )
+where
```
Comment (jwoudenberg-nri) [inline nri-log-explorer/src/Modifiable.hs]: 😻

File: nri-log-explorer/src/Main.hs [hunk 3]
```diff
@@ -72,8 +70,16 @@ data Model = Model
 
 data Filter
   = NoFilter
-  | HasFilter Text (List Text)
-  | EditFilter (Maybe (Text, List Text)) (Edit.Editor Text Name)
+  | HasFilter (Edit.Editor Text Name)
+  | EditFilter (Undo (Edit.Editor Text Name))
```
Comment (jwoudenberg-nri) [inline nri-log-explorer/src/Main.hs:74]: 😻

File: nri-log-explorer/src/Main.hs [hunk 4]
```diff
@@ -54,10 +54,8 @@ data Model = Model
     currentTime :: Time.UTCTime,
     -- A tool like pbcopy or xclip for copying to clipboard.
     clipboardCommand :: Maybe Text,
-    -- All root spans.
-    allRootSpans :: Data.Vector.Vector RootSpan,
     -- The actual data displayed.
-    filteredRootSpans :: ListWidget.List Name RootSpan,
+    rootSpans :: Modifiable.ListWidget Name RootSpan,
```
Comment (jwoudenberg-nri) [inline nri-log-explorer/src/Main.hs]: 😻 love consolidation of these two fields!

File: nri-log-explorer/src/Main.hs [hunk 5]
```diff
@@ -764,12 +758,12 @@ handleEvent pushMsg model event =
     (Brick.MouseUp _ _ _) -> Brick.continue model
     (Brick.AppEvent msg) -> update model msg
 
-data Mode = NormalMode | EditMode (Maybe (Text, List Text)) (Edit.Editor Text Name)
+data Mode = NormalMode | EditMode (Undo (Edit.Editor Text Name))
```
Comment (jwoudenberg-nri) [inline nri-log-explorer/src/Main.hs:761]: 😻 Great that this mode type has become simpler! Would be nice if we can eventually figure out a way to turn it into a plain enum I think.

**Review by jwoudenberg-nri — APPROVED** [review]
This is much nicer, thank you!

**Pull request merged**